File: src/main/Registration.java
package main;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;

import java.security.PublicKey;
import java.util.ArrayList;
import java.util.List;

//@JsonIgnoreProperties(ignoreUnknown = true)
public class Registration {
    private String id;


    private List<CourseSection> courseSections = new ArrayList<>();

    /*
    registrationStatus:
    0, means advisor has not approved.
    1, means advisor has not checked yet.
    2, means advisor has approved.
     */
    private int registrationStatus;

    public Registration() {

    }

    public String getId() {
        return id;
    }

    public void setId(String id) {
        this.id = id;
    }

    public boolean addCourseSection(CourseSection courseSectionAdded) {
        if (courseSections.size() >= 5) {
            System.out.println("WARNING ! : (Total course number exceeds 5) " + courseSectionAdded.getCourse().getId() + " - " + courseSectionAdded.getSectionNumber() + " cannot added to registration");
            return false;
        }

        if (courseSectionAdded != null) {
            for (CourseSection courseSection : courseSections) {
                if (courseSectionAdded.getCourse().getId().equals(courseSection.getCourse().getId()) && courseSectionAdded.getSectionNumber() == courseSection.getSectionNumber()) {
                    System.out.println("WARNING ! : (Same course and section exists in your registration) " + courseSectionAdded.getCourseId() + " - " + courseSectionAdded.getSectionNumber() + " cannot added to registration");
                    return false;
                }
                else if (courseSectionAdded.getCourse().getId().equals(courseSection.getCourse().getId()) && courseSectionAdded.getSectionNumber() != courseSection.getSectionNumber()) {
                    System.out.println("WARNING ! : (You try to select a second section for the same course) " + courseSectionAdded.getCourseId() + " - " + courseSectionAdded.getSectionNumber() + " cannot added to registration");
                    return false;
                }
            }

            courseSections.add(courseSectionAdded);
            System.out.println(courseSectionAdded.getCourse().getId() + " - " + courseSectionAdded.getSectionNumber() + " added to registration");
            return true;
        }
        else {
            System.out.println("WARNING ! : " + courseSectionAdded.getCourseId() + " - " + courseSectionAdded.getSectionNumber() + " cannot added to registration");
            return false;
        }

    }

    public List<CourseSection> getCourseSections() {
        return courseSections;
    }

    public void setCourseSections(List<CourseSection> courseSections) {
        this.courseSections = courseSections;
    }


    public void deleteCourseSectionList() {
        courseSections.clear();
    }

    public int getRegistrationStatus() {
        return registrationStatus;
    }

    public void setRegistrationStatus(int status) {
        registrationStatus = status;
    }


    /*Iteration 2
    public main.Student getStudent() {
        return student;
    }

     */
}

File: main/Registration.py
from typing import List

#from CourseSection import CourseSection


class Registration:
    __id : str
    __courseSections : List
    __registrationStatus : int


    def __init__(self, id = "", courseSections = [], registrationStatus = 1):
        self.__id = id

        if (len(courseSections) != 0 and type(courseSections[0]) == dict):
            for i in range(len(courseSections)):
                from CourseSection import CourseSection
                courseSections[i] = CourseSection(**courseSections[i])

            self.__courseSections = courseSections
        else:
            self.__courseSections = courseSections

        self.__registrationStatus = registrationStatus

    def to_dict(self):
        return{
            "id": self.__id,
            "courseSections": [courseSection.to_dict() if hasattr(courseSection, "to_dict") else courseSection for courseSection in self.__courseSections] if self.__courseSections else [],
            "registrationStatus": self.__registrationStatus
        }
    def getId (self):
        return self.__id

    def setId (self, id):
        self.__id = id

    def addCourseSection(self,courseSectionAdded):
        from CourseSection import CourseSection
        courseSectionAdded: CourseSection
        if(len(self.__courseSections) >= 5):
            print("WARNING ! : (Total course number exceeds 5" + courseSectionAdded.getCourse.getId()+ " - " + courseSectionAdded.getSectionNumber() + " cannot added to registration")
            return False
        if (courseSectionAdded != None) :
            for courseSection in self.__courseSections :
                if (courseSectionAdded.getCourse().getId() == (courseSection.getCourse().getId()) and courseSectionAdded.getSectionNumber() == courseSection.getSectionNumber()):
                    print("WARNING ! : (Same course and section exists in your registration) " + courseSectionAdded.getCourseId() + " - " , courseSectionAdded.getSectionNumber() , " cannot added to registration")
                    return False

                elif (courseSectionAdded.getCourse().getId() == (courseSection.getCourse().getId()) and courseSectionAdded.getSectionNumber() != courseSection.getSectionNumber()):
                    print("WARNING ! : (You try to select a second section for the same course) " + courseSectionAdded.getCourseId() + " - " , courseSectionAdded.getSectionNumber() , " cannot added to registration")
                    return False



            self.__courseSections.append(courseSectionAdded)
            print(courseSectionAdded.getCourse().getId() + " - ", courseSectionAdded.getSectionNumber() , " added to registration")
            return True

        else:
            print("WARNING ! : " + courseSectionAdded.getCourseId() + " - " , courseSectionAdded.getSectionNumber() ," cannot added to registration")
            return False


    def getCourseSections(self) :
        return self.__courseSections


    def setCourseSections(self, course_sections):
        self.__courseSections = course_sections


    def deleteCourseSectionList(self):
        self.__courseSections.clear()


    def getRegistrationStatus(self):
        return self.__registrationStatus


    def setRegistrationStatus(self, status):
        self.__registrationStatus = status